### PR TITLE
feat: usage data consent

### PR DESCRIPTION
**Pull request by bombguy**
*Issue #, if available:*

*Description of changes:*
- Added helper class to interact with config file locally
- Added text on create workflow to explain telemetry data collection
- Added subcommand under `amplify configure telemetry` to adjust user settings for telemetry.

Create amplify workflow
```
➜  amplify-gen2-app npm create amplify@latest --y
Installing required dependencies...
Creating template files...
Successfully created a new project!
Welcome to AWS Amplify! 
Run `amplify help` for a list of available commands. 
Get started by running `amplify sandbox`.
Amplify (Gen 2) collects anonymous telemetry data about general usage of the CLI.

Participation is optional, and you may opt-out by using `amplify configure telemetry disable`.

To learn more about telemetry, visit https://docs.amplify.aws/gen2/reference/telemetry
```

Enable / Disable Telemetry
```
➜  amplify-gen2-app amplify configure telemetry enable           
You have enabled telemetry data collection
➜  amplify-gen2-app ls ~/Library/Preferences/amplify
config.json
➜  amplify-gen2-app cat ~/Library/Preferences/amplify/config.json
{"telemetry":{"enabled":true,"anonymousId":"699eaae96211a5038331935fec10e82438e3aa815865f0cbbbdc785398d68abe"}}%                                                                                                        
➜  amplify-gen2-app amplify configure telemetry disable
You have disabled telemetry data collection
➜  amplify-gen2-app cat ~/Library/Preferences/amplify/config.json
{"telemetry":{"enabled":false,"anonymousId":"699eaae96211a5038331935fec10e82438e3aa815865f0cbbbdc785398d68abe"}}% 
```

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: packages/platform-core/src/config/config_controller.ts [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+      .reduce((acc: Record<string, unknown>, current: string) => {
+        return acc?.[current] as Record<string, unknown>;
+      }, this.store) as T;
+  }
+
+  /**
+   * Set value by path & update config file to disk.
+   */
+  set(path: string, value: string | boolean | number, writeToFile = true) {
+    let current: Record<string, unknown> = this._store || this.store;
```
Comment (Amplifiyer) [inline packages/platform-core/src/config/config_controller.ts:63]: We can just use `this.store`

File: packages/platform-core/src/config/config_controller.ts [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+      .split('.')
+      .reduce((acc: Record<string, unknown>, current: string) => {
+        return acc?.[current] as Record<string, unknown>;
+      }, this.store) as T;
+  }
+
+  /**
+   * Set value by path & update config file to disk.
+   */
+  set(path: string, value: string | boolean | number, writeToFile = true) {
```
Comment (Amplifiyer) [inline packages/platform-core/src/config/config_controller.ts:62]: In which situations would we not want to write to file?

File: packages/platform-core/src/config/config_controller.ts [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+   */
+  clear() {
+    this._store = {};
+    fs.rmSync(this.configFilePath);
+  }
+
+  /**
+   * Returns the path to config directory depending on OS
+   */
+  getConfigPath(name: string): string {
```
Comment (Amplifiyer) [inline packages/platform-core/src/config/config_controller.ts:103]:
```suggestion
  getConfigDirPath(name: string): string {
```

File: packages/platform-core/src/config/config_controller.ts [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+    const homedir = os.homedir();
+
+    const macos = () => path.join(homedir, 'Library', 'Preferences', name);
+    const windows = () => {
+      return path.join(
+        process.env.APPDATA || path.join(homedir, 'AppData', 'Roaming'),
+        name,
+        'Config'
```
Comment (Amplifiyer) [inline packages/platform-core/src/config/config_controller.ts:111]: Why do we need extra `Config` subdir in windows?

File: packages/platform-core/src/config/config_controller.ts [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+    };
+
+    switch (process.platform) {
+      case 'darwin':
+        return macos();
+      case 'win32':
+        return windows();
+      default:
+        return linux();
+    }
```
Comment (Amplifiyer) [inline packages/platform-core/src/config/config_controller.ts:128]: should the default be an error thrown saying we don't support this platform?

File: packages/platform-core/src/config/config_controller.ts [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+  /**
+   * Getter for cached config, retrieves config from disk if not cached already.
+   * If the store is not cached & config file does not exist, it will create a blank one.
+   */
+  private get store(): Record<string, unknown> {
+    if (this._store) {
+      return this._store;
+    }
+    if (fs.existsSync(this.configFilePath)) {
+      this._store = JSON.parse(fs.readFileSync(this.configFilePath).toString());
```
Comment (Amplifiyer) [inline packages/platform-core/src/config/config_controller.ts:33]: can we use async methods here?

File: packages/platform-core/API.md [hunk 7]
```diff
@@ -30,6 +30,22 @@ export enum CDKContextKey {
...
+    // (undocumented)
+    configFilePath: string;
+    // (undocumented)
+    dirPath: string;
+    get<T>(path: string): T;
+    getConfigPath(name: string): string;
+    set(path: string, value: string | boolean | number, writeToFile?: boolean): void;
+    // (undocumented)
+    _store: Record<string, unknown>;
+    write(): void;
```
Comment (Amplifiyer) [inline packages/platform-core/API.md:46]: These should be private. We should only have `get`, `set` and `clear`, everything else is internal detail.

File: packages/platform-core/src/config/config_controller.ts [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+  configFilePath: string;
+  _store: Record<string, unknown>;
+
+  /**
+   * Initializes common paths to telemetry configs.
+   */
+  constructor(
+    private readonly projectName = 'amplify',
+    private readonly configFileName = 'config.json'
+  ) {
```
Comment (Amplifiyer) [inline packages/platform-core/src/config/config_controller.ts:19]: Is this thread safe? e.g. if there are two instances who are doing
```ts
const inst1 = new ConfigController();
const inst2 = new ConfigController();
inst1.set('a.b', 1);
inst2.set('c.d', 1);
inst1.set('a.b', 2);
```
I'm guessing that this will wipe out `c.d` key and this is for a single threaded execution and no race condition. It would be much better if we have smaller/specific config files and a factory that vends out configController instances based on the fileName

File: packages/platform-core/src/usage-data/constants.ts [hunk 9]
```diff
@@ -7,3 +7,7 @@ export const latestApiVersion = 'v1.0';
  * returns the latest available payload version
  */
 export const latestPayloadVersion = '1.1.0';
+
+export const TELEMETRY_ENABLED_KEY = 'telemetry.enabled';
+export const TELEMETRY_ANONYMOUS_ID = 'telemetry.anonymousId';
```
Comment (Amplifiyer) [inline packages/platform-core/src/usage-data/constants.ts:12]: what's the purpose of this?

File: packages/platform-core/src/usage-data/constants.ts [hunk 10]
```diff
@@ -7,3 +7,7 @@ export const latestApiVersion = 'v1.0';
...
+export const TELEMETRY_CONFIG_KEY = 'telemetry';
```
Comment (Amplifiyer) [inline packages/platform-core/src/usage-data/constants.ts:13]: Are we using this?

File: packages/platform-core/src/usage-data/noop_usage_data_emitter.ts [hunk 11]
```diff
@@ -0,0 +1,22 @@
...
+  emitSuccess(): Promise<void> {
+    // no-op
+    return Promise.resolve();
+  }
+
+  /**
+   * no-op emitFailure
+   */
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  emitFailure(_: Error): Promise<void> {
```
Comment (sobolk) [inline packages/platform-core/src/usage-data/noop_usage_data_emitter.ts:19]:
```suggestion
  emitFailure(): Promise<void> {
```

one can skip parameters when overriding.

File: packages/platform-core/src/usage-data/usage_data_emitter_factory.ts [hunk 12]
```diff
@@ -19,6 +21,15 @@ export class UsageDataEmitterFactory {
    * Returns a ClientConfigGenerator for the given BackendIdentifier type
    */
   getInstance = (libraryVersion: string): UsageDataEmitter => {
+    const configController = new ConfigController();
+
+    if (
+      process.env.AMPLIFY_DISABLE_TELEMETRY ||
+      configController.get<boolean>(TELEMETRY_ENABLED_KEY) === false
```
Comment (sobolk) [inline packages/platform-core/src/usage-data/usage_data_emitter_factory.ts:28]: Do we need environment variable?

File: packages/platform-core/src/config/config_controller.ts (hunk 1 above)
Comment (bombguy) [reply packages/platform-core/src/config/config_controller.ts:63]: Let me re-test this piece again, suggested change was my initial implementation. This was to changed to this cover one edge case where you try to set when cache is not populated yet.

File: packages/platform-core/src/config/config_controller.ts (hunk 2 above)
Comment (bombguy) [reply packages/platform-core/src/config/config_controller.ts:62]: In case you want to batch up several updates to config to write at once.
```
controller.set('foo.bar', true, false);
controller.set('foo.baz', 12, false);
controler.write()
```

Comment (sobolk) [reply packages/platform-core/src/config/config_controller.ts:62]: For that there should be API `set(Record<string, string>)`. 
Do we have a need for that optimization?

File: packages/platform-core/src/usage-data/usage_data_emitter_factory.ts (hunk 12 above)
Comment (bombguy) [reply packages/platform-core/src/usage-data/usage_data_emitter_factory.ts:28]: The env var is primary way that the user can disable telemetry in CI environments.

File: packages/platform-core/API.md [hunk 13]
```diff
@@ -30,6 +30,22 @@ export enum CDKContextKey {
     DEPLOYMENT_TYPE = "amplify-backend-type"
 }
 
+// @public
+export class ConfigController {
```
[the remark this answers is not in the document]
Comment (bombguy) [reply packages/platform-core/API.md:34]: We can rename it to `LocalConfigurationController`

File: packages/cli/src/commands/configure/telemetry/configure_telemetry_command.ts [hunk 14]
```diff
@@ -0,0 +1,76 @@
...
+   * @inheritDoc
+   */
+  builder = (yargs: Argv) => {
+    return yargs
+      .command('enable', 'Enable anonymous data collection', {}, () => {
+        this.configController.set(TELEMETRY_ENABLED_KEY, true);
+        this.configController.set(
+          TELEMETRY_ANONYMOUS_ID,
+          this.getAnonymousId()
+        );
```
[the remark this answers is not in the document]
Comment (bombguy) [reply packages/cli/src/commands/configure/telemetry/configure_telemetry_command.ts:52]: I was hoping to use this as session ID for our telemetry. The `telemetry` object that several other CLI are generating are following similar schema. I can follow up on this one to update where the usage data emitter is emitting the metric

File: packages/platform-core/src/usage-data/constants.ts (hunk 10 above)
Comment (bombguy) [reply packages/platform-core/src/usage-data/constants.ts:13]: Yea we are using these constants to retrieve values from config in usage_data_emitter_factory & configure_telemetry.command

**Review by edwardfoyle — COMMENTED** [review]
We need to scrub the word "telemetry" from the changes in this PR. Use "usage-data" instead (This is an amazon wording policy)

File: packages/platform-core/src/usage-data/constants.ts (hunk 10 above)
Comment (Amplifiyer) [reply packages/platform-core/src/usage-data/constants.ts:13]: I meant this key `TELEMETRY_CONFIG_KEY` I couldn't find where it was being read

File: packages/platform-core/src/config/config_controller.ts (hunk 6 above)
Comment (edwardfoyle) [reply packages/platform-core/src/config/config_controller.ts:33]: We should also specify `utf-8` format to `readFile` rather than calling `.toString`

File: packages/platform-core/src/config/config_controller.ts [hunk 15]
```diff
@@ -0,0 +1,130 @@
...
+    this.dirPath = this.getConfigPath(this.projectName);
+    this.configFilePath = path.join(this.dirPath, this.configFileName);
+  }
+
+  /**
+   * Getter for cached config, retrieves config from disk if not cached already.
+   * If the store is not cached & config file does not exist, it will create a blank one.
+   */
+  private get store(): Record<string, unknown> {
```
Comment (edwardfoyle) [inline packages/platform-core/src/config/config_controller.ts:28]: `store` sounds like it should be "storing" a value. But here it is "getting the data". However, IMO we shouldn't provide a getter for the whole store anyway. It violates "tell don't ask" and it would allow some other code to modify the store without going through this class interface

File: packages/platform-core/src/config/config_controller.ts (hunk 2 above)
Comment (edwardfoyle) [reply packages/platform-core/src/config/config_controller.ts:62]: Should the path here be an enum to discourage setting random things in the store?

File: packages/platform-core/src/config/config_controller.ts [hunk 16]
```diff
@@ -0,0 +1,130 @@
...
+   * Creates project directory to store config if it doesn't exist yet.
+   */
+  private mkConfigDir() {
+    fs.mkdirSync(this.dirPath, { recursive: true });
+  }
+
+  /**
+   * Gets values from cached config by path.
+   */
+  get<T>(path: string) {
```
Comment (edwardfoyle) [inline packages/platform-core/src/config/config_controller.ts:51]: instead of just casting to T, we could pass in an "expected type" of "number" | "string" | "boolean" (or maybe just boolean if that's all we need right now) and validate that the value is the expected type?

File: packages/platform-core/src/config/config_controller.ts (hunk 8 above)
Comment (edwardfoyle) [reply packages/platform-core/src/config/config_controller.ts:19]: Another option here is to make ConfigController a singleton (should probably do that regardless)

File: packages/cli/src/commands/configure/telemetry/configure_telemetry_command.ts (hunk 14 above)
[the remark this answers is not in the document]
Comment (edwardfoyle) [reply packages/cli/src/commands/configure/telemetry/configure_telemetry_command.ts:52]: The usage data upload is already creating session IDs per command. I don't think we need to store anything besides an "enabled" but in the config file

File: packages/platform-core/src/config/config_controller.ts (hunk 1 above)
Comment (bombguy) [reply packages/platform-core/src/config/config_controller.ts:63]: looks like this works as intended and unit test still passes

File: packages/platform-core/src/config/config_controller.ts (hunk 8 above)
Comment (Amplifiyer) [reply packages/platform-core/src/config/config_controller.ts:19]: It has to be a static instance nevertheless that factory would generate, but we should still create an instance specific for a specific file instead of a general purpose config file

```ts
Factory {
  getInstance(config) {
    if(instanceExists(config)) {
      return existingInstance
    }
    create newInstance(config);
  }
}
```

File: packages/platform-core/src/config/config_controller.ts (hunk 5 above)
Comment (bombguy) [reply packages/platform-core/src/config/config_controller.ts:128]: possible values for `process.platform` according to node types are following:
"aix", "android", "darwin", "freebsd", "haiku", "linux", "openbsd", "sunos", "win32", "cygwin", "netbsd"

The other options are linux-like. Until proven otherwise, I think this is safe to assume its linux.

File: packages/platform-core/src/usage-data/constants.ts (hunk 9 above)
Comment (bombguy) [reply packages/platform-core/src/usage-data/constants.ts]: removed unnecessary keys

**Review by Amplifiyer — DISMISSED** [review]

**Comment by Amplifiyer** [comment]
Looks good, please take a look at the failing test.

**Review by srquinn21 — CHANGES_REQUESTED** [review]
There are still multiple uses of the word "Telemetry" instead of "UsageData" throughout this pull request. Please scrub as indicated by @edwardfoyle as this is a public repo and the use of "Telemetry" violates policy.

**Review by srquinn21 — APPROVED** [review]

**Review by Amplifiyer — APPROVED** [review]
